### ISSUE TITLE
bootstrap4 custom form widgets support

### ISSUE THREAD

**Issue by last-partizan** [zostera/django-bootstrap4#116]
Hello, i'm using custom-radio and custom-checkbox in my project, and to support it i had to override FieldRenderer.

https://getbootstrap.com/docs/4.1/components/forms/#checkboxes

Now something changed and i need to adjust my code, maybe it's better to implement support for those widgets in project?
If it would be useful, i could implement it and make pull-request.

**Comment by sedrubal** [comment]
I'd also like to use the custom control inputs as widgets. I especially like the [Switch from bootstra 4.2](https://getbootstrap.com/docs/4.2/components/forms/#switches).

**Comment by r03** [comment]
Support for the [custom file input](https://getbootstrap.com/docs/4.3/components/input-group/#custom-file-input) would also be nice.

**Comment by dyve** [comment]
Yes, this would be very cool indeed. The Django way would be to use the default widget unless another widget is specified. If a form has a Widget that is recognizable as a Bootstrap4Widget, the FieldRenderer could delegate the HTML rendering to the widget?

**Comment by last-partizan** [comment]
@dyve 
Probably it can be done this way, but i don't see any widget-rendering delegation inside django-bootstrap4.

Right now in my project i'm using custom field renderer with this mixin:

```
class AwesomeRenderer:
    """
    Render checkboxes for custom wodgets
    https://getbootstrap.com/docs/4.1/components/forms/#checkboxes
    """
    CUSTOM_CONTROL_CLASS = {
        "radio radio-success": "custom-control custom-radio",
        "checkbox": "custom-control custom-checkbox",
    }

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if isinstance(self.widget, forms.CheckboxInput):
            self.widget.attrs.update({"class": "custom-control-input"})

    def add_checkbox_label(self, html):
        return html + render_label(
            content=self.field.label,
            label_for=self.field.id_for_label,
            label_title=escape(strip_tags(self.field_help)),
            label_class="custom-control-label",
        )

    def wrap_widget(self, html):
        if isinstance(self.widget, forms.CheckboxInput):
            html = '<div class="custom-control custom-checkbox">{content}</div>'.format(content=html)
        return html

    def list_to_class(self, html, klass):
        klass = self.CUSTOM_CONTROL_CLASS.get(klass, klass)
        html = super().list_to_class(html, klass)
        html = html.replace('class=""', 'class="custom-control-input"')
        return html

    radio_list_to_class = list_to_class
```

And i can update FieldRenderer with same logic (using beatyfulsoup for list_to_class), and add option like "use_custom_select/use_custom_radio". (or maybe select_css_class/radio_css_class) if it makes more sense.

What do you think?

**Comment by last-partizan** [comment]
https://github.com/last-partizan/django-bootstrap4/commit/834212046106636aded8dcebdd056fcc695d2079

Here is how it looks like.

**Comment by aidanlister** [comment]
I'd like custom-select support too. I'd want the API to look like:

```{% bootstrap_field form.type show_label=False addon_after_class='' input_class='custom-select' %}```

**Comment by SebCorbin** [comment]
Hello there, I'd really like to see this implemented in `django-bootstrap4` and I'm glad someone has something functional.

Here's a list:

- [ ] Checkboxes
- [ ] Radios
- [ ] Switches
- [ ] Select
- [ ] Range
- [ ] File input

Checkboxes, Radios and Select should be fairly easy but I suggest implementing the rest through their own PR.

I like the idea to be able to set this globally, but should we handle an option like `custom_element=True` on the bootstrap_field tag? 

What's the next step after https://github.com/last-partizan/django-bootstrap4/commit/834212046106636aded8dcebdd056fcc695d2079? Would it be enough for a PR?

**Comment by dennohpeter** [comment]
> I'd like custom-select support too. I'd want the API to look like:
> 
> `{% bootstrap_field form.type show_label=False addon_after_class='' input_class='custom-select' %}`
Is this now supported ? I really need input_class paramater too

**Comment by dyve** [comment]
Not coming to django-bootstrap4, which is in maintenance mode.

django-bootstrap5 has this natively through the `checkbox_style` and `checkbox_layout` settings, which cover the custom control variants including switches.

For django-bootstrap4, the workable route stays a custom field renderer, as in the version @last-partizan linked above, registered through the `field_renderers` setting in `BOOTSTRAP4`.

Closing, with thanks to everyone who worked out the mechanics here. The checklist in this thread fed directly into how bootstrap5 handles it.
